src/kernel/include/retype_tbl.h

```diff
@@ -84,12 +115,16 @@ extern struct retype_info_glb glb_retype_tbl[N_MEM_SETS];
 
 /* get the index of the memory set. */
 #define GET_MEM_IDX(pa)                                                                 \
-	(((u32_t)pa >= COS_MEM_START) ? (((u32_t)(pa)-COS_MEM_START) / RETYPE_MEM_SIZE) \
-	                              : (((u32_t)(pa)-chal_kernel_mem_pa) / RETYPE_MEM_SIZE + N_USER_MEM_SETS))
+	(((u32_t)pa >= COS_MEM_START) ? (((u32_t)(pa) - COS_MEM_START) / RETYPE_MEM_SIZE) \
+	                              : (((u32_t)(pa) - chal_kernel_mem_pa) / RETYPE_MEM_SIZE + N_USER_MEM_SETS))
+/* The minimum/maximum page order - currently 4kB/4MB pages */
+#define MIN_PAGE_ORDER               12
+#define MAX_PAGE_ORDER               22
 /* get the memory set struct of the current cpu */
-#define GET_RETYPE_ENTRY(idx) ((&(retype_tbl[get_cpuid()].mem_set[idx])))
+#define GET_RETYPE_ENTRY(idx, order) ((&(retype_tbl[get_cpuid()].mem_set[((idx) + N_MEM_SETS) >> ((order) - MIN_PAGE_ORDER)])))
```

Comment:
This is trying to find the slot of the trunk of the memory, in the retype table. Because the retype table contains many layers now, this expression will result in the slot that corresponds to that page.